File: Objective-C/CT/Colors Trace/VC/ColorsTraceViewController.m
//
//  ViewController.m
//  CT
//
//  Created by Development on 24.06.2022.
//

#import "ColorsTraceViewController.h"
#import "ColorSelectionHandler.h"
#import "ColorCanvasHandler.h"

@interface ColorsTraceViewController ()
@property (strong, nonatomic) IBOutlet UIView *colorView;
@property (strong, nonatomic) IBOutletCollection(UIButton) NSArray *colorButtons;
@property (strong, nonatomic) IBOutlet UIBarButtonItem *undoBarButtonItem;
@property (strong, nonatomic) IBOutlet UIBarButtonItem *redoBarButtonItem;
@property (strong, nonatomic) ColorSelectionHandler *colorSelectionHandler;
@property (strong, nonatomic) ColorCanvasHandler *colorCanvasHandler;
@end

@implementation ColorsTraceViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    [self updateSelectedColorBarButtonItemByColor: nil];
            
    self.colorSelectionHandler = [[ColorSelectionHandler alloc] initWithColorButtons:self.colorButtons withColorWasChangedHandler:^(UIColor *color) {
        __weak typeof(self) weakSelf = self;
        [weakSelf updateSelectedColorBarButtonItemByColor: color];
        [weakSelf.colorCanvasHandler colorWasChangedWith: color];
    }];
    
    self.colorCanvasHandler = [[ColorCanvasHandler alloc] initWithCanvas:self.colorView withWasChangedHandler:^(BOOL isUndoEnabled, BOOL isRedoEnabled) {
        __weak typeof(self) weakSelf = self;
        [weakSelf.undoBarButtonItem setEnabled: isUndoEnabled];
        [weakSelf.redoBarButtonItem setEnabled: isRedoEnabled];
    }];
}

- (void)updateSelectedColorBarButtonItemByColor:(UIColor *) color {
    UIBarButtonItem *barButtonItem;
    
    if (color) {
        CGFloat size = 32.0;
        UIView *view = [[UIView alloc] initWithFrame: CGRectMake(0.0, 0.0, size, size)];
        view.layer.cornerRadius = size * 0.5;
        view.backgroundColor = color;
        barButtonItem = [[UIBarButtonItem alloc] initWithCustomView:view];
    }
    
    self.navigationItem.leftBarButtonItem = barButtonItem;
}

- (IBAction)buttonWasTapped:(UIButton *)sender {
    [self.colorSelectionHandler colorWasChanged: sender];
}

- (IBAction)undoPressed:(UIBarButtonItem *)sender {
    [self.colorCanvasHandler undo];
}

- (IBAction)redoPressed:(UIBarButtonItem *)sender {
    [self.colorCanvasHandler redo];
}

@end

File: SwiftUIKit/CT/Colors Trace/VC/ColorsTraceViewController.swift
//
//  ViewController.swift
//  CT
//
//  Created by Development on 23.06.2022.
//

import UIKit

class ColorsTraceViewController: UIViewController {

    @IBOutlet var colorView: UIView!
    @IBOutlet var colorButtons: [UIButton]!
    @IBOutlet var undoBarButtonItem: UIBarButtonItem!
    @IBOutlet var redoBarButtonItem: UIBarButtonItem!
        
    private lazy var colorSelectionHandler: ColorSelectionHandler? = {
        let handler = ColorSelectionHandler(colors: colorButtons) { [weak self] color in
            self?.updateSelectedColorBarButtonItem(with: color)
            self?.colorCanvasHandler?.colorWasChanged(color)
        }
        return handler
    }()
    
    private lazy var colorCanvasHandler: ColorCanvasHandler? = {
        let handler = ColorCanvasHandler(canvas: colorView) { [weak self] isUndoEnabled, isRedoEnabled in
            self?.undoBarButtonItem.isEnabled = isUndoEnabled
            self?.redoBarButtonItem.isEnabled = isRedoEnabled
        }
        return handler
    }()
           
    private func updateSelectedColorBarButtonItem(with color: UIColor? = nil) {
        var barButtonItem: UIBarButtonItem?
        
        defer {
            self.navigationItem.leftBarButtonItem = barButtonItem
        }
        
        guard color != nil else {
            return
        }
        
        let size = 32.0
        let view = UIView(frame: CGRect(x: 0.0, y: 0.0, width: size, height: size))
        view.layer.cornerRadius = size * 0.5
        view.backgroundColor = color
        
        barButtonItem = UIBarButtonItem(customView: view)
    }
        
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        self.updateSelectedColorBarButtonItem()
    }

}

extension ColorsTraceViewController {
    
    @IBAction func buttonWasTapped(_ sender: UIButton) {
        colorSelectionHandler?.wasChanged(sender)
    }
    
    @IBAction func undoPressed(_ sender: UIBarButtonItem) {
        colorCanvasHandler?.undo()
    }
    
    @IBAction func redoPressed(_ sender: UIBarButtonItem) {
        colorCanvasHandler?.redo()
    }
    
}

